ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -510,17 +518,16 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
         }
     }
     
-    if (NULLVP != virtualizationRootVNode)
+    if (VirtualizationRoot_IsValidRootHandle(rootIndex))
     {
-        vnode_put(virtualizationRootVNode);
+        vfs_setauthcache_ttl(vnode_mount(virtualizationRootVNode), 0);
     }
-    
-    if (rootIndex >= 0)
+
+    if (NULLVP != virtualizationRootVNode)
     {
-        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
-        vfs_setauthcache_ttl(vnode_mount(root->rootVNode), 0);
+        vnode_put(virtualizationRootVNode);
```

Comment:
Newbie to this area of code, I see how we have a "get" on both paths above. 

How did this work previously?  What woudl the "put" be pairing with?

Reply:
Previously, we'd skip both a put and a get by setting `virtualizationRootVNode = NULLVP;` on line 491, which has now been deleted. This particular `vnode_put()` here now always pairs with the `vnode_lookup()` at the top of the function - `vnode_lookup` returns a vnode with an iocount if it succeeds. Previously we'd normally pair that with a `vnode_put` when the provider disconnects, but there's a race condition where `vfs_setauthcache_ttl(vnode_mount(root->rootVNode), 0)` could hypothetically be called on a vnode without a refcount if the provider immediately disconnected. This way is also somewhat clearer in my opinion - and it makes the next stage, fixing the `s_virtualizationRoots` resize, more straightforward.